src/serialization/binary_archive.h

```diff
@@ -166,26 +164,27 @@ struct binary_archive<false> : public binary_archive_base<std::istream, false>
     serialize_int(t);
   }
 
-  size_t remaining_bytes() {
-    if (!stream_.good())
-      return 0;
-    //std::cerr << "tell: " << stream_.tellg() << std::endl;
-    assert(stream_.tellg() <= eof_pos_);
-    return eof_pos_ - stream_.tellg();
-  }
-
+  size_t remaining_bytes() const noexcept { return good() ? bytes_.size() : 0; }
   void enable_varint_bug_backward_compatibility() { varint_bug_backward_compatibility_ = true; }
   bool varint_bug_backward_compatibility_enabled() const { return varint_bug_backward_compatibility_; }
-
 protected:
-  std::streamoff eof_pos_;
+  epee::span<const std::uint8_t> bytes_;
+  std::uint8_t const* const begin_;
+  bool good_;
   bool varint_bug_backward_compatibility_;
 };
 
 template <>
-struct binary_archive<true> : public binary_archive_base<std::ostream, true>
+struct binary_archive<true> : public binary_archive_base<true>
 {
-  explicit binary_archive(stream_type &s) : base_type(s) { }
+  typedef std::ostream stream_type;
```

Comment:
I see, that the assumption now is, that every stream has to be compatible with `std::ostream` . Is this the case for other types of streams?

Reply:
I'm not sure what your asking - writing always required a type derived from `std::ostream`. This `typedef` was originally in `binary_archive_base` but has been moved to here.

Reply:
Understood.